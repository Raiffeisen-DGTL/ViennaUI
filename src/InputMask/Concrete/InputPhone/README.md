# InputPhone
## Импорт

```
import { InputPhone } from 'vienna-ui';
``` 
### Props

| Prop | Type | Default | Description |
| --- | --- | --- | --- |
| `prefix` | `ReactNode` | — | Значение, отображаемое перед компонентом |
| `onFocus` | `InputEventFocusEvent<HTMLInputElement, Element>> \| undefined` | — | Обработчик события при получении фокуса компонентом |
| `onBlur` | `InputEventFocusEvent<HTMLInputElement, Element>> \| undefined` | — | Обработчик события при потере фокуса компонентом |
| `ref` | `Ref<HTMLInputElement> \| undefined` | — | Ссылка на нативный элемент `<input>`, доступна после отрисовки |
| `size` | `ResponsiveProp<'s' \| 'xs' \| 'm' \| 'l' \| 'xl' \| 'xxl', Breakpoints> \| undefined` | — | Размеры компонента |
| `defaultValue` | `string \| undefined` | — | Значение в поле ввода по умолчанию (для первого рендера) |
| `autoCapitalize` | `string \| undefined` | — | Автоматическая установка заглавной буквы |
| `onCopy` | `ClipboardEventHandler<HTMLInputElement> \| undefined` | — | Обработчик события копирования |
| `onCut` | `ClipboardEventHandler<HTMLInputElement> \| undefined` | — | Обработчик события вырезки |
| `onPaste` | `ClipboardEventHandler<HTMLInputElement> \| undefined` | — | Обработчик события вставки |
| `onKeyDown` | `KeyboardEventHandler<HTMLInputElement> \| undefined` | — | Обработчик события при нажатии кнопки клавиатуры, когда компонент в фокусе |
| `onKeyPress` | `KeyboardEventHandler<HTMLInputElement> \| undefined` | — | Обработчик события при нажатии и удержании кнопки с печатаемым символом, когда компонент в фокусе |
| `onKeyUp` | `KeyboardEventHandler<HTMLInputElement> \| undefined` | — | Обработчик события при отпускании кнопки клавиатуры, когда компонент в фокусе |
| `as` | `KnownTarget \| undefined` | — | Динамически изменяет отображаемый компонент или HTML-тег, например: `as="input"` |
| `forwardedAs` | `KnownTarget \| undefined` | — | Используется для передачи другого тега/компонента |
| `theme` | `DefaultTheme \| undefined` | — | Тема компонента |
| `design` | `'outline' \| 'material' \| undefined` | — | Дизайн компонента |
| `align` | `'left' \| 'center' \| 'right' \| undefined` | — | Выравнивание текста внутри поля ввода |
| `invalid` | `boolean \| undefined` | — | Устанавливает индикацию ошибки |
| `smartPlaceholder` | `ReactNode` | — | Умное значение placeholder |
| `onDispose` | `() => void \| undefined` | — | Вызывается при удалении компонента |
| `placeholderValueAutoDiff` | `boolean \| undefined` | `true` | Автоматическое сравнение и вычитание введённого значения из `smartPlaceholder` |
| `onUpdated` | `() => void \| undefined` | — | Вызывается, когда значение обновлено |
| `private props` | — | — | — |
| `extraNativeProps` | `{ [key: string]: any; } \| undefined` | — | Для передачи дополнительных нативных атрибутов, которые могут конфликтовать с имеющимися пропсами (например, `size`) |
| `postfix` | `ReactNode` | — | Значение, отображаемое после компонента |
| `active` | `boolean \| undefined` | — | Принудительный ховер |
| `additionalPostfix` | `ReactNode` | — | Режим для включения иконки "Close" в компоненте `Search` |
| `viewOnly` | `boolean \| undefined` | — | Компонент отображается как текст, без поля ввода |
| `viewOnlyText` | `ReactNode` | — | Значение, отображаемое в режиме `viewOnly` |
| `onComplete` | `((value: string, maskRef: InputMaskFactoryOpts>, e?: InputEvent \| undefined) => void) \| undefined` | — | Вызывается, когда ввод завершён |
| `parent` | `Masked<any> \| undefined` | — | Родительский маскированный объект |
| `prepare` | `(chars: string, masked: Masked<any>, flags: AppendFlagsMaskedState>) => string \| [string, ChangeDetails] \| undefined` | — | Преобразует значение перед применением маски |
| `prepareChar` | `(chars: string, masked: Masked<any>, flags: AppendFlagsMaskedState>) => string \| [string, ChangeDetails] \| undefined` | — | Преобразует **каждый символ** перед применением маски |
| `validate` | `(value: string, masked: Masked<any>, flags: AppendFlagsMaskedState>) => boolean \| undefined` | — | Проверяет, является ли значение допустимым |
| `commit` | `(value: string, masked: Masked<any>) => void \| undefined` | — | Выполняет дополнительную логику после завершения ввода |
| `format` | `(value: DateValue, masked: Masked<any>) => string` \| `(value: number, masked: Masked<any>) => string` \| `(value: string, masked: Masked<any>) => string` \| `(value: any, masked: Masked...>) => string` \| `undefined` | — | Форматирует typed value в строку |
| `parse` | `(str: string, masked: Masked<any>) => DateValue` \| `(str: string, masked: Masked<any>) => number` \| `(str: string, masked: Masked<any>) => string` \| `(str: string, masked: Masked...>) => any` \| `undefined` | — | Парсит строку в typed value |
| `overwrite` | `boolean \| 'shift' \| undefined` | — | Включает режим перезаписи символов |
| `eager` | `boolean \| 'remove' \| 'append' \| undefined` | — | Режим "eager" для маски |
| `skipInvalid` | `boolean \| undefined` | — | Пропускать недопустимые символы |
| `autofix` | `boolean \| 'pad' \| undefined` | — | Автоматически исправляет значение |
| `definitions` | `Definitions \| undefined` | — | — |
| `blocks` | `{ [key: string]: ExtendFactoryArgOptions<BlockExtraOptions>; } \| undefined` | — | — |
| `placeholderChar` | `string \| undefined` | — | Единичный символ для отображения в пустом поле ввода |
| `displayChar` | `string \| undefined` | — | Единичный символ для заполненных позиций |
| `lazy` | `boolean \| undefined` | — | Отображает placeholder только если это необходимо |
| `mask` | `string \| RegExp \| MaskedDate \| typeof Masked \| typeof MaskedPattern \| MaskedNumber \| MaskedPattern<string> \| MaskedEnum \| ... 16 more ... \| undefined` | — | — |
| `delimiter` | `'.' \| ',' \| undefined` | — | Разделитель для чисел |
| `value` | `number \| null \| undefined` | — | — |
| `scale` | `number \| undefined` | — | Количество знаков после точки |
| `thousandsSeparator` | `string \| undefined` | — | Разделитель тысячных (один символ) |
| `padFractionalZeros` | `boolean \| undefined` | — | Флаг для добавления нулей после точки в конце ввода |
| `isFocusAllowMask` | `boolean \| undefined` | — | Позволяет фокусироваться на маске |
| `onChange` | `{ value: number \| null; event: ChangeEvent<HTMLInputElement>; options: { name?: string; isComplete: boolean; unmaskedValue?: unknown; }; } \| undefined` | — | Обработчик изменения значения |
| `mapToRadix` | `string[] \| undefined` | — | Массив единичных символов |
| `radix` | `string \| undefined` | — | Единичный символ для разделения целой и дробной части |
| `normalizeZeros` | `boolean \| undefined` | — | Флаг для удаления ведущих и завершающих нулей в конце ввода |
| `enum` | `string[] \| undefined` | — | Массив строк для enum-маски |
| `matchValue` | `(enumStr: string, inputStr: string, matchFrom: number) => boolean \| undefined` | — | Проверяет совпадение значения enum и ввода |
| `from` | `number \| undefined` | — | Минимальное значение диапазона |
| `to` | `number \| undefined` | — | Максимальное значение диапазона |
| `dispatch` | `(appended: string, masked: MaskedDynamic<any>, flags: AppendFlagsHandleState>, tail: string \| String \| TailDetails) => Masked...> \| undefined` | — | Выбирает {@link Masked} в зависимости от введённого значения |

---

###  HTMLAttributes

| Prop | Type | Default | Description |
| --- | --- | --- | --- |
| `autoComplete` | `string \| undefined` | — | Включение встроенного в браузер автокомплита |
| `disabled` | `boolean \| undefined` | — | Компонент неактивен, если `true` |
| `maxLength` | `number \| undefined` | — | Максимальная длина вводимого значения в символах. Опционально устанавливает максимальную длину шаблона. Используется, когда длина шаблона превышает `to`. Добавляет нули в начало в этом случае. |
| `name` | `string \| undefined` | — | Имя компонента |
| `placeholder` | `string \| undefined` | — | Значение, отображаемое, если ничего не введено |
| `readOnly` | `boolean \| undefined` | — | Компонент активен, но не доступен для редактирования |
| `type` | `string \| undefined` | — | Типы поля ввода |
| `autoFocus` | `boolean \| undefined` | — | Автофокус при монтировании |
| `className` | `string \| undefined` | — | Название стиля для компонента (опционально) |
| `id` | `string \| undefined` | — | Уникальный идентификатор компонента |
| `spellCheck` | `boolean \| undefined` | — | Если `true`, работает системная проверка правописания |
| `tabIndex` | `number \| undefined` | `-1` | Порядок получения фокуса при нажатии `Tab` |
| `autoCorrect` | `string \| undefined` | — | Включение встроенного в браузер корректирования ввода |




# InputPhone

Компонент ввода номера телефона



```
    <InputPhone value={'4959191242'} />
```

## Внешний вид

Дизайн, размеры и другие настройки полностью идентичны компоненту `Input`. Для использования необходимо ознакомится с работой `IMask`. Подробнее о свойствах смотрите в `InputMask`;

```
    {() => {
        const [value, setValue] = React.useState('');
        const handleChange = React.useCallback(({value}) => setValue(value), []);
        return (
            <Groups>
                <InputPhone value={value} onChange={handleChange} size='xs' />
                <InputPhone value={value} onChange={handleChange} size='s' />
                <InputPhone value={value} onChange={handleChange} size='l' />
                <InputPhone value={value} onChange={handleChange} size='xl' />
            </Groups>
        );
    }}
```

По умолчанию маска настроена для России.

```
    <InputPhone value={'4959191242'} disabled />
```

По умолчанию включена передача через `8` вместо `+7`

```
    <InputPhone value={'84959191242'} disabled />
```

## Пример с редактированием поля

```
    {() => {
        const [value, setValue] = React.useState('');
        const handleChange = React.useCallback(({ value }) => setValue(value), []);
        return <InputPhone value={value} onChange={handleChange} />;
    }}
```

## Вариант с предустановленой частью номера

```
    {() => {
        const [value, setValue] = React.useState('');
        const handleChange = React.useCallback(({ value }) => setValue(value), []);
        return <InputPhone value={value} mask={'+{7} (495) 000-00-00'} onChange={handleChange} />;
    }}
```

## Вариант со сложной маской телефона

```
    {() => {
        const [value, setValue] = React.useState('');
        const handleChange = React.useCallback(({ value }) => setValue(value), []);
        return <InputPhone value={value} mask={'+{42} (000) {234}-00-00'} smartPlaceholder="+42 ___ ___-__-__" placeholder="+42 ___ ___-__-__" onChange={handleChange} />;
    }}
```

## Вариант с передачей кастомного замещающего символа

```
    {() => {
        const [value, setValue] = React.useState('');
        const handleChange = React.useCallback(({ value }) => setValue(value), []);
        return <InputPhone value={value} mask={'+{42} (000) {234}-00-00'} smartPlaceholder="+42 ___ ___-__-__"  placeholder="+42 ___ ___-__-__" onChange={handleChange} />;
    }}
```

## Состояние ViewOnly

Это состояние используется, когда нужно показать значение поля без возможности изменения.
Может использоваться для построения форм, которые находятся в режиме просмотра, где все поля заполнены, но не доступны для редактирования.

Свойства:

- viewOnly - состояние `ViewOnly` (тип boolean);
- viewOnlyText - текст значения (тип ReactNode);

```
    <InputPhone viewOnly value={'+7 213 213-12-31'} />
```