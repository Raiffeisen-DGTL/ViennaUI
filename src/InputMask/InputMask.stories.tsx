import React, {useCallback, useState } from 'react';
import { Story, Meta } from 'storybook';
import { InputMask, InputMaskProps } from './InputMask';

export default {
    title: 'Development/InputMask',
    component: InputMask,
} as Meta;

export const Overview: Story<InputMaskProps> = (args) => {
    const [value, setValue] = useState('');
    const changeHandler = useCallback((value) => setValue(value), []);
    return <InputMask maskOptions={{mask: Date}} value={value} onChange={changeHandler} {...args} />;
};
