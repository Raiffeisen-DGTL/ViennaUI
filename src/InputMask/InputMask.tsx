import React, { Ref, forwardRef, useEffect } from 'react';
import { useIMask, ReactMaskProps } from 'react-imask';
import { FactoryOpts, InputElement } from 'imask';
import { Input, InputProps } from '../Input';

export type InputMaskProps<Opts extends FactoryOpts = FactoryOpts> = Omit<InputProps, 'value' | 'onChange'> &
    Partial<Pick<ReactMaskProps<InputElement, Opts>, 'onComplete'>> & {
        maskOptions: Opts;
        value?: unknown;
        onChange?: (value: unknown) => void;
    };

export const InputMask = forwardRef<HTMLInputElement, InputMaskProps>(
    ({ value: externalValue, maskOptions, onChange, onComplete, ...rest }, externalRef) => {
        const onAccept = (maskedValue, maskRef, inputEvent) => {
            // вызываем onChange только на изменение инпута пользователем(в остальных случаях событие не приходит)
            if (inputEvent) {
                onChange?.(maskedValue);
            }
        };

        const { ref, value, setValue, setTypedValue, maskRef } = useIMask<HTMLInputElement, FactoryOpts>(maskOptions, {
            onAccept,
            onComplete,
        });

        useEffect(() => {
            if (typeof externalValue === 'string') {
                setValue(externalValue);
            } else {
                setTypedValue(externalValue);
            }
        }, [externalValue]);

        useEffect(() => {
            maskRef.current?.updateValue();
            if (externalValue === '') {
                setValue('');
            }
        }, [value]);

        return <Input value={value} {...rest} ref={composeRef(ref, externalRef)} />;
    }
);

InputMask.displayName = 'InputMask';
InputMask.defaultProps = {
    size: 'l',
    design: 'outline',
};

export default InputMask;

function composeRef<T>(...refs: Ref<T>[]): Ref<T> {
    if (refs.length <= 1) {
        return refs[0];
    }

    return (node: T) => {
        refs.forEach((ref) => {
            if (typeof ref === 'function') {
                ref(node);
            } else if (typeof ref === 'object' && ref && 'current' in ref) {
                (ref as unknown as Record<string, T>).current = node;
            }
        });
    };
}
