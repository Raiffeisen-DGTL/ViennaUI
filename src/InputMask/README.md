# InputMask

Компонент ввода текстовых данных по маске.

## Импорт

```
import { InputMask } from 'vienna-ui';
```

## Свойства / Props

| Prop | Type | Default | Description |
| --- | --- | --- | --- |
| onChange | (InputEvent<FormEvent<HTMLInputElement>> & ((value: unknown) => void)) \| undefined |  |
| design | "outline" \| "material" \| undefined |  | Дизайн |
| prefix | ReactNode |  | Значанеие отображаемое перед компонентом |
| autoCapitalize | string \| undefined |  | Автоматическая установка заглавной буквы |
| onCopy | ClipboardEventHandler  \| undefined |  | Обработчик копирования |
| onCut | ClipboardEventHandler  \| undefined |  | Обработчик вырезки |
| onPaste | ClipboardEventHandler \| undefined |  | Обработчик вставки |
| onFocus | InputEvent<FocusEvent<HTMLInputElement, Element>> \| undefined |  | Обработчик события при получении фокуса компонентом |
| onBlur | InputEvent<FocusEvent<HTMLInputElement, Element>> \| undefined |  | Обработчик события при потере фокуса компонентом |
| onKeyDown | KeyboardEventHandler \| undefined |  | Обработчик события при нажатии кнопки клавиатуры, когда компонент в фокусе |
| onKeyPress | KeyboardEventHandler \| undefined |  | Обработчик события при нажатии и удержании кнопки клавиатуры с печатемым символом, когда компонент в фокусе |
| onKeyUp | KeyboardEventHandler \| undefined |  | Обработчик события при отпускании кнопки клавиатуры, когда компонент в фокусе |
| onMouseDown | Function \| undefined |  | Обработчик события первого полупериода клика |
| onPointerDown | Function \| undefined |  | Обработчик события первого полупериода клика |
| size | ResponsiveProp<"xs" \| "s" \| "m" \| "l" \| "xl" \| "xxl", Breakpoints> \| undefined |  | Размеры |
| value | string \| undefined |  | Значение в поле ввода |
| defaultValue | string \| undefined |  |
| ref | Ref<HTMLInputElement>  \| undefined |  | Сcылка на нативный элемент input, доступна после отрисовки |
| invalid | boolean \| undefined |  |
| placeholderValueAutoDiff | boolean \| undefined |  |
| active | boolean \| undefined |  | Принудительный ховер |
| postfix | ReactNode |  | Значение отображаемое после компонента |
| smartPlaceholder | ReactNode |  |
| onDespose | (() => void) \| undefined |  |
| onUpdated | (() => void) \| undefined |  |
| onComplete | ((value: string, maskRef: InputMask<FactoryOpts>, e?: InputEvent \| undefined) => void) \| undefined |  |
| maskOptions | FactoryOpts |



## HTMLAttributes

| Prop | Type | Default | Description |
| --- | --- | --- | --- |
| id | string \| undefined | |
| title | string \| undefined | |
| className | string \| undefined | |
| spellCheck | boolean \| undefined | |
| tabIndex | number \| undefined | |
| height | string \| undefined | |
| width | string \| undefined | |
| style | CSSProperties \| undefined | |
| name | string \| undefined | |
| type | string \| undefined | |
| maxLength | number \| undefined | |
| autoComplete | string \| undefined | |
| autoCorrect | string \| undefined | |
| autoFocus | boolean \| undefined | |
| readOnly | boolean \| undefined | |
| placeholder | string \| undefined | |
| disabled | boolean \| undefined | |

# InputMask

Компонент ввода данных по маске


```
    {() => {
        const [value, setValue] = React.useState('');
        const changeHandler = React.useCallback((e, data) => {
            setValue(data.value);
        }, []);
        return (
            <InputMask
                value={value}
                onChange={changeHandler}
                maskOptions={{mask: Date}}
                placeholder='Введите дату в формате ДД.ММ.ГГГГ'
            />
        );
    }}
```

## Внешний вид

Дизайн, размеры и другие настройки полностью идентичны компоненту Input. Для использования необходимо ознакомится с работой IMask.

```
    {() => {
        const [value, setValue] = React.useState('');
        const changeHandler = React.useCallback((e, data) => setValue(data.value), []);
        return (
            <Groups design='vertical'>
                <InputMask
                    value={value}
                    onChange={changeHandler}
                    maskOptions={{mask: Date}}
                    placeholder='Введите дату в формате ДД.ММ.ГГГГ'
                />
            </Groups>
        );
    }}
```

## Стандартные элементы

Для удобства реализовано несколько стандартных компонентов маски.

##### Account

```
    <InputAccount value='111111' />
```

##### Card

```
    <InputCard value='111111' />
```

##### Дата

```
    <InputDate value='22.02.1' />
```

##### Дата с периодом

```
    <InputDateRange value='11.12.1900 - 12' />
```

##### Digital

```
    <InputDigital value='1231' />
```

##### Число

```
    <InputNumber value='123123123.123' />
```

##### Телефон

```
    <InputPhone value='749539' />
```

## Паттерны масок

Чтобы задать паттерн для маски, нужно использовать:
0 - для любого числа,
a - для любой буквы, * для любого символа

Чтобы 0, а или * использовались как обычные символы, их нужно экранировать с помощью двойного обратного слэша

```
    {() => {
        const [value, setValue] = React.useState('');
        const changeHandler = React.useCallback((data) => setValue(data), []);
        return (
            <Groups design='vertical'>
                <InputMask
                    value={value}
                    onChange={changeHandler}
                    maskOptions={{mask:"\\0000"}}
                />
            </Groups>
        );
    }}
```
