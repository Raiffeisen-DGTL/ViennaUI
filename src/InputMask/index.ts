export { InputMask, type InputMaskProps } from './InputMask';
export { InputPhone, type InputPhoneProps } from './Concrete/InputPhone/InputPhone';
export { InputAccount, type InputAccountProps } from './Concrete/InputAccount/InputAccount';
export { InputNumber, type InputNumberProps } from './Concrete/InputNumber/InputNumber';
export { InputDate, type InputDateProps, type InputDateLocalization } from './Concrete/InputDate';
export { InputCard, type InputCardProps } from './Concrete/InputCard/InputCard';
export {
    InputDateRange,
    type InputDateRangeProps,
    type DateType,
    type InputDateRangeLocalization,
} from './Concrete/InputDateRange';
export { InputDigital, type InputDigitalProps } from './Concrete/InputDigital/InputDigital';
