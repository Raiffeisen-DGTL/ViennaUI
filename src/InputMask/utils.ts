// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
import { AllFactoryStaticOpts, FactoryOpts } from 'imask';
import { format } from 'date-fns';

export const getMaskOptionsFromProps = (props: FactoryOpts) => {
    const keys: (keyof AllFactoryStaticOpts)[] = [
        'mask',
        'parent',
        'prepare',
        'prepareChar',
        'placeholderChar',
        'validate',
        'commit',
        'format',
        'parse',
        'overwrite',
        'radix',
        'mapToRadix',
        'eager',
        'min',
        'max',
        'skipInvalid',
    ];

    const options = keys.reduce((res, key) => {
        if (props[key] !== undefined) {
            res[key] = props[key];
        }

        return res;
    }, {}) as Partial<AllFactoryStaticOpts>;

    return options;
};

const isValidDate = <T>(d: T): boolean => {
    return d instanceof Date && !Number.isNaN(d.getTime());
};

export const dateFormat = (value: string | Date) => {
    return isValidDate(value) ? format(value as Date, 'dd.MM.yyyy') : typeof value === 'string' ? value : '';
};
