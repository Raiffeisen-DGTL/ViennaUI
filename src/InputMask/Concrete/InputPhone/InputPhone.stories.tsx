import React, { useCallback, useState } from 'react';
import { Story, Meta } from 'storybook';
import { InputPhone } from './InputPhone';
import { InputMaskProps } from '../../InputMask';

export default {
    title: 'Development/InputPhone',
    component: InputPhone,
} as Meta;

export const Overview: Story<any> = () => {
    const [value, setValue] = useState();
    const changeHandler = useCallback(({ value: data }) => setValue(data), []);

    return <InputPhone onChange={changeHandler} value={value} />;
};

export const DifferentMask: Story<InputMaskProps> = () => {
    const [value, setValue] = React.useState('');
    const handleChange = React.useCallback(({ value: data }) => {
        setValue(data);
    }, []);

    return (
        <InputPhone
            value={value}
            mask={'+{42} (000) {234}-00-00'}
            placeholderChar={'#'}
            smartPlaceholder={'+42 (###) 234-##-##'}
            placeholder={'+42 (###) 234-##-##'}
            onChange={handleChange}
        />
    );
};

DifferentMask.storyName = 'Вариант с передачей кастомного замещающего символа';

export const MaskWithSetTimeout: Story<InputMaskProps> = () => {
    const [value, setValue] = React.useState('');
    const handleChange = React.useCallback(({ value: data }) => setValue(data), []);

    React.useEffect(() => {
        setTimeout(() => {
            setValue('+7 213 213-12-31');
        }, 2000);
    }, []);

    return <InputPhone value={value} onChange={handleChange} />;
};

MaskWithSetTimeout.storyName = 'Вариант с setTimeout';

export const ViewOnly: Story<any> = () => {
    return <InputPhone viewOnly value={'+7 213 213-12-31'} />;
};
