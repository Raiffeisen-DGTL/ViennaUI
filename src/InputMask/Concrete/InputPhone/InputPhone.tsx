import React, { forwardRef, useState, useEffect } from 'react';
import { FactoryOpts } from 'imask';
import { InputMask, InputMaskOnChangeType, InputMaskProps } from '../../InputMask';
import { getMaskOptionsFromProps } from '../../utils';

export type InputPhoneProps = Omit<InputMaskProps, 'value' | 'maskOptions' | 'onChange'> &
    FactoryOpts & {
        value?: string;
        onChange?: InputMaskOnChangeType<string>;
    };

const defaultPlaceholder = '+7 000 000-00-00';

export const InputPhone = forwardRef<HTMLInputElement, InputPhoneProps>((props, ref) => {
    const {
        onFocus,
        onBlur,
        onKeyDown,
        placeholder = defaultPlaceholder,
        smartPlaceholder = defaultPlaceholder,
        value: externalValue,
        onChange,
        ...attrs
    } = props;
    const [value, setValue] = useState(externalValue);

    useEffect(() => {
        setValue(externalValue);
    }, [externalValue]);

    const handleFocus: InputMaskProps['onFocus'] = (e, data) => {
        if (typeof onFocus === 'function') {
            onFocus(e, data);
        }
    };

    const handleBlur: InputMaskProps['onBlur'] = (e, data) => {
        if (typeof onBlur === 'function') {
            onBlur(e, data);
        }
    };

    const handleKeyDown: InputMaskProps['onKeyDown'] = (e) => {
        if (typeof onKeyDown === 'function') {
            onKeyDown(e);
        }
    };

    const handleChange: InputMaskProps['onChange'] = ({ value: maskedValue, event, options }) => {
        if (typeof maskedValue !== 'string') return;
        setValue(maskedValue);
        if (typeof onChange === 'function') {
            onChange({ value: maskedValue, event, options });
        }
    };

    return (
        <InputMask
            ref={ref}
            value={value || ''}
            maskOptions={{
                mask: '+{7} 000 000-00-00',
                placeholderChar: '',
                prepare: (char, { value }) => {
                    if (!value) {
                        if (char.startsWith('8')) {
                            return `7${char.substring(1)}`;
                        }
                        if (char.startsWith('+7')) {
                            return `7${char.substring(2)}`;
                        }
                    }

                    return char;
                },
                ...getMaskOptionsFromProps(props),
            }}
            placeholder={placeholder}
            smartPlaceholder={smartPlaceholder}
            onKeyDown={handleKeyDown}
            onFocus={handleFocus}
            onBlur={handleBlur}
            onChange={handleChange}
            {...attrs}
        />
    );
});

InputPhone.displayName = 'InputPhone';
