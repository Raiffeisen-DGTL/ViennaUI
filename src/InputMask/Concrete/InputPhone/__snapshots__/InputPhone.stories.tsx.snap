// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Development/InputPhone DifferentMask smoke-test 1`] = `
<style>
  @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Regular.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Regular.woff') format('woff');
        font-weight: normal;
        font-style: normal;
    }

    @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Medium.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Medium.woff') format('woff');
        font-weight: 500;
        font-style: normal;
    }

    @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Bold.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Bold.woff') format('woff');
        font-weight: bold;
        font-style: normal;
    }

    @font-face {
        font-family: 'Navigo';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/Navigo-Regular-Web.woff') format('woff');
        font-weight: normal;
        font-style: normal;
    }

    @font-face {
        font-family: 'Navigo';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/Navigo-Bold-Web.woff') format('woff');
        font-weight: bold;
        font-style: normal;
    }
</style>
<div class="sc-bcPKhP psOLg">
  <div class="sc-jrkPcm juzgae">
    <label class="sc-iBAaJG eDDvAp">
      <div class="sc-gKHVLF hxumYj">
        <div class="sc-dkjKgF dCerbJ">
          <input mask="+{42} (000) {234}-00-00"
                 placeholderchar="#"
                 aria-invalid="false"
                 placeholder="+42 (###) 234-##-##"
                 class="sc-hLclGa hqQfvS"
          >
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`Development/InputPhone MaskWithSetTimeout smoke-test 1`] = `
<style>
  @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Regular.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Regular.woff') format('woff');
        font-weight: normal;
        font-style: normal;
    }

    @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Medium.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Medium.woff') format('woff');
        font-weight: 500;
        font-style: normal;
    }

    @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Bold.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Bold.woff') format('woff');
        font-weight: bold;
        font-style: normal;
    }

    @font-face {
        font-family: 'Navigo';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/Navigo-Regular-Web.woff') format('woff');
        font-weight: normal;
        font-style: normal;
    }

    @font-face {
        font-family: 'Navigo';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/Navigo-Bold-Web.woff') format('woff');
        font-weight: bold;
        font-style: normal;
    }
</style>
<div class="sc-bcPKhP psOLg">
  <div class="sc-jrkPcm juzgae">
    <label class="sc-iBAaJG eDDvAp">
      <div class="sc-gKHVLF hxumYj">
        <div class="sc-dkjKgF dCerbJ">
          <div class="sc-ftLKQv bIkLsL">
            <div class="sc-grXZZQ hSkQrD">
              <span class="sc-pqitP bCKaoL">
                +7 213 213-12-31
              </span>
              <span>
              </span>
            </div>
          </div>
          <input aria-invalid="false"
                 placeholder="+7 000 000-00-00"
                 class="sc-hLclGa hqQfvS"
          >
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`Development/InputPhone Overview smoke-test 1`] = `
<style>
  @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Regular.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Regular.woff') format('woff');
        font-weight: normal;
        font-style: normal;
    }

    @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Medium.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Medium.woff') format('woff');
        font-weight: 500;
        font-style: normal;
    }

    @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Bold.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Bold.woff') format('woff');
        font-weight: bold;
        font-style: normal;
    }

    @font-face {
        font-family: 'Navigo';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/Navigo-Regular-Web.woff') format('woff');
        font-weight: normal;
        font-style: normal;
    }

    @font-face {
        font-family: 'Navigo';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/Navigo-Bold-Web.woff') format('woff');
        font-weight: bold;
        font-style: normal;
    }
</style>
<div class="sc-bcPKhP psOLg">
  <div class="sc-jrkPcm juzgae">
    <label class="sc-iBAaJG eDDvAp">
      <div class="sc-gKHVLF hxumYj">
        <div class="sc-dkjKgF dCerbJ">
          <input aria-invalid="false"
                 placeholder="+7 000 000-00-00"
                 class="sc-hLclGa hqQfvS"
          >
        </div>
      </div>
    </label>
  </div>
</div>
`;
